File: python/samples/02-agents/tools/code_interpreter.py
# Copyright (c) Microsoft. All rights reserved.

import asyncio

from agent_framework import ChatAgent, HostedCodeInterpreterTool
from agent_framework.openai import OpenAIResponsesClient

"""
Code Interpreter Tool

Demonstrates using HostedCodeInterpreterTool to let the agent write and execute
Python code for calculations, data analysis, and problem solving.

The code interpreter runs in a sandboxed environment managed by the provider.

For more on code interpreter:
- File downloads: getting_started/agents/openai/openai_responses_client_with_code_interpreter_files.py
- Azure AI: getting_started/agents/azure_ai/azure_ai_with_code_interpreter.py
- Docs: https://learn.microsoft.com/agent-framework/concepts/code-interpreter
"""


async def main() -> None:
    print("=== Code Interpreter Tool ===\n")

    # <create_agent>
    agent = ChatAgent(
        chat_client=OpenAIResponsesClient(),
        instructions="You are a helpful assistant that can write and execute Python code to solve problems.",
        tools=HostedCodeInterpreterTool(),
    )
    # </create_agent>

    # <run_query>
    query = "Calculate the factorial of 100"
    print(f"User: {query}")
    result = await agent.run(query)
    print(f"Result: {result}\n")
    # </run_query>

    # <inspect_code>
    # Inspect the generated code and outputs
    for message in result.messages:
        code_blocks = [c for c in message.contents if c.type == "code_interpreter_tool_call"]
        outputs = [c for c in message.contents if c.type == "code_interpreter_tool_result"]
        if code_blocks:
            code_inputs = code_blocks[0].inputs or []
            for content in code_inputs:
                if content.type == "text":
                    print(f"Generated code:\n{content.text}")
                    break
        if outputs:
            print("Execution outputs:")
            for out in outputs[0].outputs or []:
                if out.type == "text":
                    print(out.text)
    # </inspect_code>


if __name__ == "__main__":
    asyncio.run(main())

File: dotnet/samples/02-agents/tools/CodeInterpreter.cs
// Copyright (c) Microsoft. All rights reserved.

// Code Interpreter Tool
// Use the hosted Code Interpreter tool for math, data analysis, and code execution.
// Shows both MEAI-based and native SDK approaches.
//
// For docs: https://learn.microsoft.com/agent-framework/agents/tools

using System.Text;
using Azure.AI.Projects;
using Azure.AI.Projects.OpenAI;
using Azure.Identity;
using Microsoft.Agents.AI;
using Microsoft.Extensions.AI;
using OpenAI.Responses;

string endpoint = Environment.GetEnvironmentVariable("AZURE_FOUNDRY_PROJECT_ENDPOINT")
    ?? throw new InvalidOperationException("AZURE_FOUNDRY_PROJECT_ENDPOINT is not set.");
string deploymentName = Environment.GetEnvironmentVariable("AZURE_FOUNDRY_PROJECT_DEPLOYMENT_NAME") ?? "gpt-4o-mini";

AIProjectClient aiProjectClient = new(new Uri(endpoint), new AzureCliCredential());

// <code_interpreter>
AIAgent agent = await aiProjectClient.CreateAIAgentAsync(
    model: deploymentName,
    name: "CoderAgent",
    instructions: "You are a personal math tutor. When asked a math question, write and run code using the python tool to answer the question.",
    tools: [new HostedCodeInterpreterTool() { Inputs = [] }]);

AgentResponse response = await agent.RunAsync("I need to solve the equation sin(x) + x^2 = 42");
// </code_interpreter>

// <read_code_output>
CodeInterpreterToolCallContent? toolCallContent = response.Messages
    .SelectMany(m => m.Contents).OfType<CodeInterpreterToolCallContent>().FirstOrDefault();
if (toolCallContent?.Inputs is not null)
{
    DataContent? codeInput = toolCallContent.Inputs.OfType<DataContent>().FirstOrDefault();
    if (codeInput?.HasTopLevelMediaType("text") ?? false)
    {
        Console.WriteLine($"Code Input: {Encoding.UTF8.GetString(codeInput.Data.ToArray())}");
    }
}

CodeInterpreterToolResultContent? toolResultContent = response.Messages
    .SelectMany(m => m.Contents).OfType<CodeInterpreterToolResultContent>().FirstOrDefault();
if (toolResultContent?.Outputs?.OfType<TextContent>().FirstOrDefault() is { } resultOutput)
{
    Console.WriteLine($"Code Tool Result: {resultOutput.Text}");
}
// </read_code_output>

await aiProjectClient.Agents.DeleteAgentAsync(agent.Name);
